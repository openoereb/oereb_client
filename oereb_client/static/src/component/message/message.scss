@import "../../oereb_client.scss";

.toast {
  border: 0;
}