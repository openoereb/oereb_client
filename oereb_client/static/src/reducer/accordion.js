import {createSlice} from "@reduxjs/toolkit";

export const accordionSlice = createSlice({
  name: 'accordion',
  initialState: {
    category: null,
    topic: null,
    viewServices: []
  },
  reducers: {
    setActiveCategory: (state, action) => {
      state.category = action.payload;
    },
    setActiveTopic: (state, action) => {
      state.topic = action.payload;
    },
    setViewServices: (state, action) => {
      state.viewServices = action.payload;
    }
  }
});

export const {setActiveCategory, setActiveTopic, setViewServices} = accordionSlice.actions;

export default accordionSlice.reducer;
