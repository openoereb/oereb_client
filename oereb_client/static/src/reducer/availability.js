import {createSlice} from "@reduxjs/toolkit";
import {isBoolean, isString} from "lodash";

export const availabilitySlice = createSlice({
    name: 'availability',
    initialState: {
        key: 'OerebAvailability',
        visible: true
    },
    reducers: {
        setAvailabilityPrefix: (state, action) => {
            state.key = action.payload + 'OerebAvailability';
        },
        initAvailability: (state) => {
            const availability = localStorage.getItem(state.key);
            if (isBoolean(availability)) {
                state.visible = availability;
            }
            else if (isString(availability)) {
                state.visible = JSON.parse(availability);
            }
            else {
                state.visible = true;
            }
        },
        showAvailability: (state, action) => {
            state.visible = action.payload;
            localStorage.setItem(state.key, state.visible);
        }
    }
});

export const {setAvailabilityPrefix, initAvailability, showAvailability} = availabilitySlice.actions;

export default availabilitySlice.reducer;
