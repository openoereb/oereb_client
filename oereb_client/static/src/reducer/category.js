import { createSlice } from "@reduxjs/toolkit";

export const categorySlice = createSlice({
    name: 'category',
    initialState: {
        category: null
    },
    reducers: {
        setActiveCategory: (state, action) => {
            state.category = action.payload;
        }
    }
});

export const { setActiveCategory } = categorySlice.actions;

export default categorySlice.reducer;
