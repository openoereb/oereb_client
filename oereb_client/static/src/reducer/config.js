import {createSlice} from "@reduxjs/toolkit";

export const configSlice = createSlice({
  name: 'config',
  initialState: {
    config: {}
  },
  reducers: {
    update: (state, action) => {
      state.config = action.payload;
    }
  }
});

export const {update} = configSlice.actions;

export default configSlice.reducer;
