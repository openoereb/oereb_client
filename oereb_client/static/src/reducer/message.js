import {createSlice} from "@reduxjs/toolkit";
import {v4 as uuidv4} from "uuid";

export const MESSAGE_TIMEOUT = 10000;

export const messageSlice = createSlice({
  name: "message",
  initialState: {
    messages: []
  },
  reducers: {
    info: (state, action) => {
      state.messages.push({
        id: uuidv4(),
        type: "info",
        text: action.payload.text,
        confirmation: action.payload.confirmation,
        timestamp: Date.now()
      });
    },
    warning: (state, action) => {
      state.messages.push({
        id: uuidv4(),
        type: "warning",
        text: action.payload.text,
        confirmation: action.payload.confirmation,
        timestamp: Date.now()
      });
    },
    error: (state, action) => {
      state.messages.push({
        id: uuidv4(),
        type: "error",
        text: action.payload.text,
        confirmation: action.payload.confirmation,
        timestamp: Date.now()
      });
    },
    close: (state, action) => {
      for (let i = 0; i < state.messages.length; i++) {
        if (state.messages[i].id === action.payload) {
          state.messages.splice(i, 1);
          break;
        }
      }
    },
    cleanMessages: (state) => {
      for (let i = state.messages.length; i > 0; i--) {
        const now = Date.now();
        if (!state.messages[i-1].confirmation && now - state.messages[i-1].timestamp >= MESSAGE_TIMEOUT) {
          state.messages.splice(i-1, 1);
        }
      }
    }
  }
});

export const {cleanMessages, info, error, warning, close} = messageSlice.actions;

export const showInfo = function(text, confirmation) {
  return function(dispatch) {
    dispatch(info({
      text: text,
      confirmation: confirmation || false
    }));
    setTimeout(() => {
      dispatch(cleanMessages());
    }, MESSAGE_TIMEOUT);
  };
}

export const showWarning = function(text, confirmation) {
  return function(dispatch) {
    dispatch(warning({
      text: text,
      confirmation: confirmation || false
    }));
    setTimeout(() => {
      dispatch(cleanMessages());
    }, MESSAGE_TIMEOUT);
  };
}

export const showError = function(text, confirmation) {
  return function(dispatch) {
    dispatch(error({
      text: text,
      confirmation: confirmation || false
    }));
    setTimeout(() => {
      dispatch(cleanMessages());
    }, MESSAGE_TIMEOUT);
  };
}

export default messageSlice.reducer;
