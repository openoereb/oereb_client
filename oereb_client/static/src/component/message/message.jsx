import { Toast } from "bootstrap";
import React, { useEffect, useRef } from "react";
import { useDispatch, useSelector } from "react-redux";
import { close, MESSAGE_TIMEOUT } from "../../reducer/message";
import "./message.scss";
import { useTranslation } from "react-i18next";
import { isString } from "lodash";

/**
 * This component shows messages (warnings, errors) in the user interface.
 */
const OerebMessage = function () {
  const element = useRef(null);
  const messages = useSelector((state) => state.message).messages;
  const language = useSelector((state) => state.language);
  const dispatch = useDispatch();
  const {t} = useTranslation();

  const closeMessage = function(messageId) {
    const toasts = document.body.querySelectorAll(".toast");
    toasts.forEach((toastEl) => {
      if (toastEl.getAttribute("data-oereb-message-id") === messageId) {
        Toast.getOrCreateInstance(toastEl).hide();
      }
    });
    setTimeout(() => {
      dispatch(close(messageId));
    }, 500);
  };

  const items = messages.map((message, key) => {
    let toastTitle;
    let toastIcon;
    if (message.type === "warning") {
      toastIcon = <i className="bi bi-exclamation-triangle-fill text-warning me-2"></i>;
      toastTitle = t("message.warning");
    }
    else if (message.type === "error") {
      toastIcon = <i className="bi bi-x-octagon-fill text-danger me-2"></i>;
      toastTitle = t("message.error");
    }
    else {
      toastIcon = <i className="bi bi-info-circle-fill text-primary me-2"></i>;
      toastTitle = t("message.info");
    }
    let text;
    if (isString(message.text)) {
      text = message.text;
    }
    else {
      text = message.text[language.current] || message.text[language.default];
    }
    return (
      <div
        key={message.id}
        className="toast"
        role="alert"
        aria-live="assertive"
        aria-atomic="true"
        data-bs-autohide="false"
        data-oereb-message-id={message.id}
        data-oereb-message-confirm={message.confirmation}
      >
        <div className="toast-header">
          {toastIcon}
          <strong className="me-auto">{toastTitle}</strong>
          <button
            type="button"
            className="btn-close"
            aria-label="Close"
            onClick={closeMessage.bind(this, message.id)}
          ></button>
        </div>
        <div className="toast-body">
          {text}
        </div>
      </div>
    );
  });

  useEffect(() => {
    const toasts = element.current.querySelectorAll(".toast");
    toasts.forEach((toastEl) => {
      if (!toastEl.hasAttribute("initialized")) {
        const toast = Toast.getOrCreateInstance(toastEl);
        toast.show();
        if (toastEl.getAttribute("data-oereb-message-confirm") !== "true") {
          setTimeout(() => {
            toast.hide();
          }, MESSAGE_TIMEOUT - 500);
        }
      }
      toastEl.setAttribute("initialized", true);
    });
  }, [items]);

  return (
    <div ref={element} className="toast-container position-absolute bottom-0 end-0 mb-3 me-3">
      {items}
    </div>
  );
};

OerebMessage.propTypes = {};

export default OerebMessage;
