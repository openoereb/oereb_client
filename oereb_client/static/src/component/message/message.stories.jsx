import React from "react";

import OerebMessage from "./message";
import { showError, showWarning } from "../../reducer/message";
import { useDispatch } from "react-redux";

export default {
  title: "API Reference/Component/Message",
  component: OerebMessage,
  tags: ["autodocs"]
};

export const Message = () => {
  const dispatch = useDispatch();

  const warning = () => {
    dispatch(showWarning("This is a warning message example."));
  };

  const error = () => {
    dispatch(showError("This is an error message example."));
  };

  return (
    <div style={{"min-height": "300px"}}>
      <button type="button" className="btn btn-primary me-3" onClick={warning}>Show warning</button>
      <button type="button" className="btn btn-primary me-3" onClick={error}>Show error</button>
      <OerebMessage />
    </div>
  );
};