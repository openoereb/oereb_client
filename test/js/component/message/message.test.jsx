import {act, render} from "@testing-library/react";
import React from "react";
import {Provider} from "react-redux";

import OerebMessage from "../../../../oereb_client/static/src/component/message/message";
import MainStore from "../../../../oereb_client/static/src/store/main";
import { error, info, warning } from "../../../../oereb_client/static/src/reducer/message";

describe("message component", () => {

  let component;

  beforeEach(() => {
    vi.mock('uuid', () => ({ v4: () => 'a81dad33-08cf-4b50-b0c6-d14bcc427df8' }));
    component = render(
      <Provider store={MainStore}>
        <OerebMessage />
      </Provider>
    );
  });

  it("should render message container", () => {
    expect(component.asFragment()).toMatchSnapshot();
  });

  it("should render info message", () => {
    act(() => {
      MainStore.dispatch(info({
        text: "foo",
        confirmation: false
      }));
    });
    expect(component.asFragment()).toMatchSnapshot();
  });

  it("should render warning message", () => {
    act(() => {
      MainStore.dispatch(warning({
        text: "foo",
        confirmation: false
      }));
    });
    expect(component.asFragment()).toMatchSnapshot();
  });

  it("should render error message", () => {
    act(() => {
      MainStore.dispatch(error({
        text: "foo",
        confirmation: false
      }));
    });
    expect(component.asFragment()).toMatchSnapshot();
  });

});