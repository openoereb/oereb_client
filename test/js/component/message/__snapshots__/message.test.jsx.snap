// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`message component > should render error message 1`] = `
<DocumentFragment>
  <div
    class="toast-container position-absolute bottom-0 end-0 mb-3 me-3"
  >
    <div
      aria-atomic="true"
      aria-live="assertive"
      class="toast fade show showing"
      data-bs-autohide="false"
      data-oereb-message-confirm="false"
      data-oereb-message-id="a81dad33-08cf-4b50-b0c6-d14bcc427df8"
      initialized="true"
      role="alert"
    >
      <div
        class="toast-header"
      >
        <i
          class="bi bi-info-circle-fill text-primary me-2"
        />
        <strong
          class="me-auto"
        >
          message.info
        </strong>
        <button
          aria-label="Close"
          class="btn-close"
          type="button"
        />
      </div>
      <div
        class="toast-body"
      >
        foo
      </div>
    </div>
    <div
      aria-atomic="true"
      aria-live="assertive"
      class="toast fade show showing"
      data-bs-autohide="false"
      data-oereb-message-confirm="false"
      data-oereb-message-id="a81dad33-08cf-4b50-b0c6-d14bcc427df8"
      initialized="true"
      role="alert"
    >
      <div
        class="toast-header"
      >
        <i
          class="bi bi-exclamation-triangle-fill text-warning me-2"
        />
        <strong
          class="me-auto"
        >
          message.warning
        </strong>
        <button
          aria-label="Close"
          class="btn-close"
          type="button"
        />
      </div>
      <div
        class="toast-body"
      >
        foo
      </div>
    </div>
    <div
      aria-atomic="true"
      aria-live="assertive"
      class="toast fade show showing"
      data-bs-autohide="false"
      data-oereb-message-confirm="false"
      data-oereb-message-id="a81dad33-08cf-4b50-b0c6-d14bcc427df8"
      initialized="true"
      role="alert"
    >
      <div
        class="toast-header"
      >
        <i
          class="bi bi-x-octagon-fill text-danger me-2"
        />
        <strong
          class="me-auto"
        >
          message.error
        </strong>
        <button
          aria-label="Close"
          class="btn-close"
          type="button"
        />
      </div>
      <div
        class="toast-body"
      >
        foo
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`message component > should render info message 1`] = `
<DocumentFragment>
  <div
    class="toast-container position-absolute bottom-0 end-0 mb-3 me-3"
  >
    <div
      aria-atomic="true"
      aria-live="assertive"
      class="toast fade show showing"
      data-bs-autohide="false"
      data-oereb-message-confirm="false"
      data-oereb-message-id="a81dad33-08cf-4b50-b0c6-d14bcc427df8"
      initialized="true"
      role="alert"
    >
      <div
        class="toast-header"
      >
        <i
          class="bi bi-info-circle-fill text-primary me-2"
        />
        <strong
          class="me-auto"
        >
          message.info
        </strong>
        <button
          aria-label="Close"
          class="btn-close"
          type="button"
        />
      </div>
      <div
        class="toast-body"
      >
        foo
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`message component > should render message container 1`] = `
<DocumentFragment>
  <div
    class="toast-container position-absolute bottom-0 end-0 mb-3 me-3"
  />
</DocumentFragment>
`;

exports[`message component > should render warning message 1`] = `
<DocumentFragment>
  <div
    class="toast-container position-absolute bottom-0 end-0 mb-3 me-3"
  >
    <div
      aria-atomic="true"
      aria-live="assertive"
      class="toast fade show showing"
      data-bs-autohide="false"
      data-oereb-message-confirm="false"
      data-oereb-message-id="a81dad33-08cf-4b50-b0c6-d14bcc427df8"
      initialized="true"
      role="alert"
    >
      <div
        class="toast-header"
      >
        <i
          class="bi bi-info-circle-fill text-primary me-2"
        />
        <strong
          class="me-auto"
        >
          message.info
        </strong>
        <button
          aria-label="Close"
          class="btn-close"
          type="button"
        />
      </div>
      <div
        class="toast-body"
      >
        foo
      </div>
    </div>
    <div
      aria-atomic="true"
      aria-live="assertive"
      class="toast fade show showing"
      data-bs-autohide="false"
      data-oereb-message-confirm="false"
      data-oereb-message-id="a81dad33-08cf-4b50-b0c6-d14bcc427df8"
      initialized="true"
      role="alert"
    >
      <div
        class="toast-header"
      >
        <i
          class="bi bi-exclamation-triangle-fill text-warning me-2"
        />
        <strong
          class="me-auto"
        >
          message.warning
        </strong>
        <button
          aria-label="Close"
          class="btn-close"
          type="button"
        />
      </div>
      <div
        class="toast-body"
      >
        foo
      </div>
    </div>
  </div>
</DocumentFragment>
`;
